### PR TITLE
Feature/8 generation filter

### LINKED ISSUE #8: Generation Filter
Create the generations filter based on the figma design and make it work with the list of pokemon on the main page.

Generation Filter button

Image

Opens the Generation filter popup

Image

When expanded

Image

### Acceptance Criteria

- [x] Get the icon Image from the style guide and add it to the header of the page.
- [x] When pressing the filter by generation button, a pop-up should be displayed like the picture above.
- [x] The pop-up should have a title, a short description and a collection of the generations shown in a list of 2 columns.
- [x] When any generation is pressed, the pokemon list should filter the pokemon by that generation.
- [x] If there's a generation selected, it should be marked as red.
- [x] When a generation is de-selected, the filter should be removed on the pokemon list and the pop-up should close.
- [x] When a generation is selected, the pop-up should close and the filter should be applied.
- [x] When the pop-up is opened for the first time, it should open on its collapsed form
- [x] When the pop-up is expanded, it should show the complete panel.
- [x] When closing the pop-up is like cancelling the action, nothing should change and the pop-up should dissapear.
- [x] Unit tests

### Epic Link
[Pokemon List Epic](https://github.com/jmacboy/Pokedex-ios/wiki/Epic-for-Pokemon-list)

### PR DESCRIPTION
**Pull request by Rodrigo-Schar**
-added generation filter
-fixed apply filter
-added unit test
-refactor for pull request
-refactor code for pull request
-merge develop and add color to constantVariables
-fixed testResources

### CODE REVIEW

File: Pokedex_ios/Assets.xcassets/IconButtons/generationFilterIcon.imageset/Contents.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "images" : [
     {
-      "filename" : "water.svg",
+      "filename" : "Vector.svg",
```
Comment (jmacboy) [inline Pokedex_ios/Assets.xcassets/IconButtons/generationFilterIcon.imageset/Contents.json]: try using a better name for your assets

File: Pokedex_ios/Helpers/ConstantVariables.swift [hunk 2]
```diff
@@ -31,6 +31,16 @@ class ConstantVariables {
         TypeElement(type: Generation(id: 17, name: "dark")),
         TypeElement(type: Generation(id: 18, name: "fairy"))
     ]
+    static let pokemonGenerations = [
+        GenerationAux(name: "Generation I", description: "generation-i"),
```
Comment (jmacboy) [inline Pokedex_ios/Helpers/ConstantVariables.swift]: why aux?

File: Pokedex_ios/View/Controller/GenerationFilterPopup/GenerationFilterPopupViewController.swift [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        containerView.addSubview(contentStackView)
+        contentStackView.translatesAutoresizingMaskIntoConstraints = false
+        NSLayoutConstraint.activate([
+            contentStackView.topAnchor.constraint(equalTo: containerView.topAnchor, constant: 32),
+            contentStackView.leadingAnchor.constraint(equalTo: containerView.leadingAnchor, constant: 40),
+            contentStackView.trailingAnchor.constraint(equalTo: containerView.trailingAnchor, constant: -40)
+        ])
+    }
+
+    func setUpElements() {
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/GenerationFilterPopup/GenerationFilterPopupViewController.swift]: setup is a single word

File: Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift [hunk 4]
```diff
@@ -26,26 +26,39 @@ class PokemonListViewController: UIViewController {
         setBackground()
         initViewModel()
         setSearchView()
-        // Setup for the advanced filter image and button
-        setUpAdvanceFilterPopup()
+        // Setup for all filter image and button
+        setFiltersPopup()
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift]: better use your own method, leave alone setupAdvanceFilterPopup and make one that is called setupGenerationPopup. Mostly for better code reading purposes

File: Pokedex_iosTests/GenerationFilterTests.swift [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    override func setUpWithError() throws {
+        mockTest = MockPokedexService()
+        viewModel = PokemonListViewModel(pokedexService: MockPokedexService())
+    }
+
+    override func tearDownWithError() throws {
+        
+    }
+
+    func testPerformanceExample() throws {
```
Comment (jmacboy) [inline Pokedex_iosTests/GenerationFilterTests.swift]: does this do something? if not remove it, also with the teardown method

**Review by jmacboy — CHANGES_REQUESTED** [review]

File: Pokedex_ios/Helpers/ConstantVariables.swift (hunk 2 above)
Comment (Rodrigo-Schar) [reply Pokedex_ios/Helpers/ConstantVariables.swift]: because i only use it for the filter

Comment (jmacboy) [reply Pokedex_ios/Helpers/ConstantVariables.swift]: use a more descriptive name, for example GenerationForPopup (I know is too long but it makes it readable once you check variables)

Comment (Rodrigo-Schar) [reply Pokedex_ios/Helpers/ConstantVariables.swift]: Maybe i should call it GenerationFilter?

Comment (Rodrigo-Schar) [reply Pokedex_ios/Helpers/ConstantVariables.swift]: ok i'll check it

Comment (jmacboy) [reply Pokedex_ios/Helpers/ConstantVariables.swift]: sure, that name could work

File: Pokedex_ios/Assets.xcassets/IconButtons/generationFilterIcon.imageset/Contents.json (hunk 1 above)
Comment (Rodrigo-Schar) [reply Pokedex_ios/Assets.xcassets/IconButtons/generationFilterIcon.imageset/Contents.json]: done!

File: Pokedex_ios/Helpers/ConstantVariables.swift (hunk 2 above)
Comment (Rodrigo-Schar) [reply Pokedex_ios/Helpers/ConstantVariables.swift]: done!

File: Pokedex_ios/View/Controller/GenerationFilterPopup/GenerationFilterPopupViewController.swift (hunk 3 above)
Comment (Rodrigo-Schar) [reply Pokedex_ios/View/Controller/GenerationFilterPopup/GenerationFilterPopupViewController.swift]: done!

File: Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift (hunk 4 above)
Comment (Rodrigo-Schar) [reply Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift]: i can't add the icons to the navigationItem separately but i changed the name of the method to be more specific

File: Pokedex_iosTests/GenerationFilterTests.swift (hunk 5 above)
Comment (Rodrigo-Schar) [reply Pokedex_iosTests/GenerationFilterTests.swift]: done!

File: Pokedex_ios/View/Cells/GenerationCell/GenerationCollectionViewCell.swift [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        super.awakeFromNib()
+    }
+
+    func setupData(generation: GenerationFilter, isTypeSelected: Bool) {
+        generationLabel.text = generation.name
+        let imageName = "pokemon-\(generation.description)"
+        generationImageView.image = UIImage(named: imageName)
+
+        if isTypeSelected {
+            generationLabel.textColor = UIColor(named: "TextWhiteColor")
```
Comment (jmacboy) [inline Pokedex_ios/View/Cells/GenerationCell/GenerationCollectionViewCell.swift]: These colors have already been declared on other parts of the app, please reuse them

File: Pokedex_ios/View/Controller/GenerationFilterPopup/GenerationFilterPopupViewController.swift [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            contentStackView.leadingAnchor.constraint(equalTo: containerView.leadingAnchor, constant: 40),
+            contentStackView.trailingAnchor.constraint(equalTo: containerView.trailingAnchor, constant: -40)
+        ])
+    }
+
+    func setupElements() {
+        self.generationCollectionView.delegate = self
+        self.generationCollectionView.dataSource = self
+
+        let uiNibPokemonGenerationCell = UINib(nibName: "GenerationCollectionViewCell", bundle: nil)
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/GenerationFilterPopup/GenerationFilterPopupViewController.swift]: Use the nibName as a constant

File: Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift [hunk 8]
```diff
@@ -26,26 +26,39 @@ class PokemonListViewController: UIViewController {
...
         pokemonsTableView.register(uiNib, forCellReuseIdentifier: pokeCellIdentifier)
     }
-    func setUpAdvanceFilterPopup() {
+    func setFiltersIconsPopup() {
         let advancedFilterImage = UIImage(named: "AdvancedFilterSVG")?.withRenderingMode(.alwaysOriginal)
         let advancedFilterButton = UIBarButtonItem(image: advancedFilterImage, style: .plain,
                                                    target: self, action: #selector(showAdvancedFilterPopup))
-        navigationItem.rightBarButtonItem = advancedFilterButton
+
+        let generationFilterImage = UIImage(named: "generationFilterIcon")?.withRenderingMode(.alwaysOriginal)
```
Comment (jmacboy) [inline Pokedex_ios/View/Controller/PokemonList/PokemonListViewController.swift:41]: The problem wasnt the method name here, it was better that you create your own method to do your stuff, not reuse the one that it was

**Review by jmacboy — CHANGES_REQUESTED** [review]

File: Pokedex_ios/View/Cells/GenerationCell/GenerationCollectionViewCell.swift [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    func setupData(generation: GenerationFilter, isTypeSelected: Bool) {
+        generationLabel.text = generation.name
+        let imageName = "pokemon-\(generation.description)"
+        generationImageView.image = UIImage(named: imageName)
+
+        if isTypeSelected {
+            generationLabel.textColor = UIColor(named: "background-white")
+            patternImageView.backgroundColor = UIColor(named: "apply-button")
+            generationBackground.backgroundColor = UIColor(named: "apply-button")
+            generationImageView.tintColor = UIColor(named: "apply-button")
```
Comment (jmacboy) [inline Pokedex_ios/View/Cells/GenerationCell/GenerationCollectionViewCell.swift]: also those strings are already defined as constants on the constants class, please reference that class.

**Review by jmacboy — CHANGES_REQUESTED** [review]
Thats the last thing and we’re done

File: Pokedex_iosTests/TestResources.swift [hunk 10]
```diff
@@ -35,23 +35,33 @@ class TestResources {
     let filterHeightsCriteria: [PokemonHeigths] = [.large]
 
     static let pokemonRaw = [
+<<<<<<< HEAD
```
Comment (jmacboy) [inline Pokedex_iosTests/TestResources.swift]: don't commit this never

File: Pokedex_ios/View/Cells/GenerationCell/GenerationCollectionViewCell.swift (hunk 9 above)
Comment (Rodrigo-Schar) [reply Pokedex_ios/View/Cells/GenerationCell/GenerationCollectionViewCell.swift]: done!

**Review by jmacboy — APPROVED** [review]

File: Pokedex_iosTests/TestResources.swift (hunk 10 above)
Comment (Rodrigo-Schar) [reply Pokedex_iosTests/TestResources.swift]: done! , i solved that conflict but i didn't check it again

**Pull request merged**